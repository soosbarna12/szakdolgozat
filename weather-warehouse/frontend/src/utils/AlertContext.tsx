import { Alert, Snackbar } from "@mui/material";
import React, { createContext, ReactNode, useContext, useState } from "react";
import { AlertContextType } from "../types/alertContext.type";


export const AlertContext = createContext<AlertContextType>({
  showAlert: () => { },
});


export function AlertProvider({ children }: { children: ReactNode }) {
  const [open, setOpen] = useState(false);
  const [message, setMessage] = useState("");
  const [severity, setSeverity] = useState<"error" | "info" | "success" | "warning">("info");

  function showAlert(newMessage: string, newSeverity: "error" | "info" | "success" | "warning" = "info") {
    setMessage(newMessage);
    setSeverity(newSeverity);
    setOpen(true);
  }

  return (
    <AlertContext.Provider value={{ showAlert }}>
      {children}
      <Snackbar open={open} autoHideDuration={4000} onClose={() => setOpen(false)}>
        <Alert severity={severity} onClose={() => setOpen(false)}>
          {message}
        </Alert>
      </Snackbar>
    </AlertContext.Provider>
  );
}

export const useAlert = () => useContext(AlertContext);
