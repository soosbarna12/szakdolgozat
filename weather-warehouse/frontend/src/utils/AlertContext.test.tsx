import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { AlertProvider, useAlert } from './AlertContext';
import '@testing-library/jest-dom';

function TestComponent() {
  const { showAlert } = useAlert();

  return (
    <div>
      <button onClick={() => showAlert('Test message', 'success')}>Show Success Alert</button>
      <button onClick={() => showAlert('Error message', 'error')}>Show Error Alert</button>
    </div>
  );
}

describe('utils/AlertContext', () => {
  it('displays the alert with the correct message and severity (success)', async () => {
    render(
      <AlertProvider>
        <TestComponent />
      </AlertProvider>
    );

    const button = screen.getByText('Show Success Alert');
    fireEvent.click(button);

    const alert = await screen.findByText('Test message');
    expect(alert).toBeInTheDocument();
    expect(alert).toMatchSnapshot();
  });

  it('displays the alert with the correct message and severity (error)', async () => {
    render(
      <AlertProvider>
        <TestComponent />
      </AlertProvider>
    );

    const button = screen.getByText('Show Error Alert');
    fireEvent.click(button);

    const alert = await screen.findByText('Error message');
    expect(alert).toBeInTheDocument();
    expect(alert).toMatchSnapshot();
  });

  it('closes the alert after the auto-hide duration', async () => {
    render(
      <AlertProvider>
        <TestComponent />
      </AlertProvider>
    );

    const button = screen.getByText('Show Success Alert');
    fireEvent.click(button);

    const alert = await screen.findByText('Test message');
    expect(alert).toBeInTheDocument();

    await waitFor(() => {
      expect(alert).not.toBeInTheDocument();
    }, { timeout: 5000 }); // Allow some buffer for the auto-hide duration
  });
});