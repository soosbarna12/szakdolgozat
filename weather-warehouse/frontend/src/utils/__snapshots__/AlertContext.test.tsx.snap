// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/AlertContext displays the alert with the correct message and severity (error) 1`] = `
<div
  class="MuiAlert-message css-zioonp-MuiAlert-message"
>
  Error message
</div>
`;

exports[`utils/AlertContext displays the alert with the correct message and severity (success) 1`] = `
<div
  class="MuiAlert-message css-zioonp-MuiAlert-message"
>
  Test message
</div>
`;
